### DISCUSSION TITLE
Modules Help - Umbrella / O365

### DISCUSSION THREAD

**Discussion by Git-Me-Some-Hub** [Security-Onion-Solutions/securityonion#4813] [Unsupported Versions]
I am currently using the ISO 2.3.60 that I just upgraded from 2.3.51. 

I am using an external Filebeats agent to ship Cisco: Umbrella, FTD, ASA, Meraki, O365, Syslog: CEF log types. I updated because I knew you all were incorporating the new easier module use, and because I was having ingest problems coming from Umbrella, FTD, and Meraki. 

I tried to input the Cisco FTD module and it seem to allow Filebeats to start up, but I am having issues with both Umbrella and O365. 
Error: Rendering SLS 'minions.sensor' failed. Please see master log for details.
- Not even sure where to look for this log, that ones I did, did not have anything. 

Test Umbrella: 

filebeat:
  third_party_filebeat:
    modules:
      cisco:
        umbrella:
          enabled: true
          var.input:aws-s3
          var.queue_url: https://sqs.us-west-1.amazonaws.com/files
          var.access_key_id: <id>
          var.secret_access_key: <key>
          var.visibility_timeout: 300s
          var.api_timeout: 120s
          var.tz_offset: local

Test O365: 

filebeat:
  third_party_filebeat:
    module:
      o365:
        audit:
          enabled: true
          var.application_id: "<appid>"
          var.tenants:
           - id: "<id>" 
             name: "domain"
          var.content_type:
	       - "Audit.AzureActiveDirectory"
	       - "Audit.Exchange"
	       - "Audit.SharePoint"
	       - "Audit.General"
	       - "DLP.All"
          var.client_secret: "<secret>"

Test FTD

filebeat:
  third_party_filebeat:
    module:
      ftd:
        enabled: true
        var.input: syslog
        var.syslog_host: localhost
        var.syslog_port: 9003

Any help on what I am doing wrong with 2 of these, but got right for the FTD...

**Comment by Git-Me-Some-Hub** [comment]
So I found the default module file and tried to keep the exact same format, and I can get it working for most cisco modules... but Umbrella. It works when I have the default and `enabled:false` but not when I put in the minimum to pull from an S3 bucket. 

Working: 
![image](https://user-images.githubusercontent.com/17930434/125826050-12890e8d-e4dd-4985-871f-e4e58eff16aa.png)

Not working
![image](https://user-images.githubusercontent.com/17930434/125826705-073d9c84-472b-4662-8638-e8198d17c5c5.png)


Even when Fielbeats restarts correctly, I do no see any change in the modules.d folder (/opt/so/conf/filebeat/etc/modules.d/).

**Comment by weslambert** [comment]
You might want to double-check your config.  It looks like you are using `module` instead of `modules`.

Ex.

```
filebeat:
  third_party_filebeat:
    modules:
```

**Reply by Git-Me-Some-Hub** [reply]
Thanks Wes, I did see that yesterday and did try it, but to no-avail. Can you clue me into the correct log file that might shed some light on this? Also I should see the config changes on that specific node in the /opt/so/conf/filebeat/etc/modules.d/ folder, correct? 
![image](https://user-images.githubusercontent.com/17930434/125985661-1db4ee6b-6fce-43a6-98ef-956920eb7c2a.png)

**Reply by weslambert** [reply]
Have you already modified filebeat.yml before updating? Could you have modified it and put it in the `/opt/so/saltstack/local/salt/filebeat/etc/` directory and not removed it after updating? 

You should be making your changes in a minion pillar file, then the config should be pushed to `/opt/so/conf/filebeat/modules` to the relevant node.

What is the output of the following from the node where you are attempting to pull in these logs via Filebeat?

`docker exec -it so-filebeat /usr/share/filebeat/filebeat modules list`

You could also try setting `logging.level` to `debug` in filebeat.yml (copying from default to local directory)

**Reply by Git-Me-Some-Hub** [reply]
I know you have had a few other major issues as I see in the discussions, but any other direction on how I could troubleshoot this would be appreciated. 

Thanks.

**Comment by Git-Me-Some-Hub** [comment]
1. I am not seeing any files after in the Filebeats folder on either my sensor that I am trying to send the config down to, or the manager. So nothing should be left over.  
2. I am doing the change in nano /opt/so/saltstack/local/pillar/minions/sensor.sls
3. When I run that command, while having that error, the container does not actually start,.
`docker exec -it so-filebeat /usr/share/filebeat/filebeat modules list`
`Error: No such container: so-filebeat`

What start / restarting so-filebeats: 
`local:
    Data failed to compile:
----------
    Pillar failed to render with the following messages:
----------
    Rendering SLS 'minions.enaa-sec07p_sensor' failed. Please see master log for details.
`
5. When I do run that docker command I am seeing all modules listed under the Disabled heading, nothing on Enabled, even though I do have Cisco ASA, FTD, and Meraki listed as enabled. 

Does that help? If it is not starting I am assuming the debug level would not help, is that right?

**Reply by weslambert** [reply]
It is assumed FIlebeat is already running when running that docker command.  If Filebeat is not starting, try tailing the log in /opt/so/log/filebeat for clues.

**Reply by Git-Me-Some-Hub** [reply]
I  see errors, but nothing that seem specific to the log files. 

Errors: - /opt/so/log/filebeat
`Sensor 
2021-07-15T01:31:05.995Z        WARN    beater/filebeat.go:178  Filebeat is unable to load the Ingest Node pipelines for the configured modules because the Elasticsearch output is not configured/enabled. If you have already loaded  the Ingest Node pipelines or are using Logstash pipelines, you can ignore this warning.
2021-07-20T18:25:11.438Z        ERROR   [publisher_pipeline_output]     pipeline/output.go:154  Failed to connect to backoff(async(tcp://managersearch:5644)): dial tcp 10.20.0.203:5644: connect: connection refused
2021-07-20T18:26:49.355Z        ERROR   [logstash]      logstash/async.go:280   Failed to publish events caused by: write tcp 172.17.0.10:58422->10.20.0.203:5644: write: connection reset by peer
2021-07-20T18:26:51.086Z        ERROR   [publisher_pipeline_output]     pipeline/output.go:180  failed to publish events: write tcp 172.17.0.10:58422->10.20.0.203:5644: write: connection reset by peer

ManagerSearch:
2021-07-20T18:25:13.714Z        ERROR   [publisher_pipeline_output]     pipeline/output.go:154  Failed to connect to backoff(async(tcp://managersearch:5644)): read tcp 172.17.0.18:37024->10.20.0.203:5644: read: connection reset by peer
2021-07-20T18:29:42.578Z        ERROR   [logstash]      logstash/async.go:280   Failed to publish events caused by: EOF
2021-07-20T18:29:42.581Z        ERROR   [logstash]      logstash/async.go:280   Failed to publish events caused by: client is not connected`

Maybe it is formatting thing? I keep Umbrella False and just add the highlighted and it stops the Filebeat Container from loading. No Filebeat logs happen during this time. Are there some Files that I can start defaulting to confirm... I am at a loss, as it seems the whole module system is not working for me. The other few enabled items never do get written to the configs to enable them. 
![image](https://user-images.githubusercontent.com/17930434/126390878-3c2f541c-91fb-42e5-949d-e99846df0de0.png)
 
I am copying the exact text form the Documentation. So-Filebeat starts up, but it does not change any configuration. I have also changed it between modules, and module. The config file: /opt/so/saltstack/default/salt/filebeat/etc/module_config.yml.jinja lists module, with no S, but again, I am not 100% sure how all these interoperate. 
![image](https://user-images.githubusercontent.com/17930434/126393492-142d19da-9680-455b-8c8f-8b6dbc69c39d.png)

Assuming that it is working for you and most people, is there an easy way that I can put some data in there, then look at a specific file to see the changes? 
On my Sensor: 
- Nothing is changed in: /opt/so/conf/filebeat/etc/modules.d/
- No files or folder in: /opt/so/saltstack/local/
- Should be no changes in: /opt/so/conf/filebeat/etc/filebeat.yml

I am really at a loss.

**Reply by Git-Me-Some-Hub** [reply]
Two follow ups 
- I updated to the latest 2.3.60 as I was planning on using Curator and turn on Elastic Auth but had a failure. 
- 
![image](https://user-images.githubusercontent.com/17930434/126569874-c963401c-d366-48e6-bf25-ec42d42b551c.png)

I then was looking for a SO-* command to see what version I was on, but saw this command: so-filebeat-module-setup
![image](https://user-images.githubusercontent.com/17930434/126569947-87cc7553-234b-400a-aba8-86f61a9b2750.png)

Anyway this is related? I do not remember seeing in the documentation anything around needing to launch the setup script.

**Reply by Git-Me-Some-Hub** [reply]
Also seeing so-dockerregistry as Missing when I use so-status, and no commands under the so-* options to start it or do anything with it. But this is probably a separate issue.

**Reply by Git-Me-Some-Hub** [reply]
Just checking back on this. Any way you think the /usr/share/filebeat/module-setup.yml file not being present is possibly the issue?

**Reply by Git-Me-Some-Hub** [reply]
Ok, I have gotten to the bottom of the issues with so-filebeat not starting. I am dumb. I knew that I needed to line up the columns spacing-wise, but I did not know **tabs were not allowed**. I had 1 or two in there, so **it is now starting so-filbeat**. 

However,  I am not seeing the right modules folder listed on my production cluster. 

Production cluster has : /opt/so/conf/filebeat/ - etc/      -registry/
: /opt/so/conf/filebeat/etc/ -filebeat.yml  -module/       -modules.d/    -pki/
- To me this shows me the updates did not work? 
- soversion: 2.3.60
- sohotfix: ECSFIX HEAVYNODE_SSL_LOGSTASH_REDIS_PIPELINES FBPIPELINE CURATORAUTH

Test cluster has: /opt/so/conf/filebeat/modules/  - securityonion.yml  -thirdparty.yml
- This is Test cluster that I rebuilt today from ISO 2.3.51 to 2.3.61 works fine. 
- soversion: 2.3.61
- sohotfix: STENODOCKER

I would run soup to see if that would help, but I have another post out there about my so-dockerregistry in a missing/pending state. https://github.com/Security-Onion-Solutions/securityonion/discussions/4901

Any manual ways around it?

**Reply by weslambert** [reply]
What does /usr/sbin/so-filebeat-module-setup look like?

**Reply by Git-Me-Some-Hub** [reply]
Here you go: 

```
default_conf_dir=/opt/so/conf
ELASTICSEARCH_HOST="10.20.0.203"
ELASTICSEARCH_PORT=9200
#ELASTICSEARCH_AUTH=""

# Define a default directory to load pipelines from
FB_MODULE_YML="/usr/share/filebeat/module-setup.yml"


# Wait for ElasticSearch to initialize
echo -n "Waiting for ElasticSearch..."
COUNT=0
ELASTICSEARCH_CONNECTED="no"
while [[ "$COUNT" -le 240 ]]; do
        curl -k --output /dev/null --silent --head --fail -L https://"$ELASTICSEARCH_HOST":"$ELASTICSEARCH_PORT"
        if [ $? -eq 0 ]; then
                ELASTICSEARCH_CONNECTED="yes"
                echo "connected!"
                break
        else
                ((COUNT+=1))
                sleep 1
                echo -n "."
        fi
done
if [ "$ELASTICSEARCH_CONNECTED" == "no" ]; then
        echo
        echo -e "Connection attempt timed out.  Unable to connect to ElasticSearch.  \nPlease try: \n  -checking log(s) in /var/log/elasticsearch/\n  -running 'sudo docker ps' \n  -running 'sudo so-elastic-restart'"
        echo
fi
echo "Testing to see if the pipelines are already applied"
ESVER=$(curl -sk https://"$ELASTICSEARCH_HOST":"$ELASTICSEARCH_PORT" |jq .version.number |tr -d \")
PIPELINES=$(curl -sk https://"$ELASTICSEARCH_HOST":"$ELASTICSEARCH_PORT"/_ingest/pipeline/filebeat-$ESVER-suricata-eve-pipeline | jq . | wc -c)

if [[ "$PIPELINES" -lt 5 ]]; then
  echo "Setting up ingest pipeline(s)"

  for MODULE in activemq apache auditd aws azure barracuda bluecoat cef checkpoint cisco coredns crowdstrike cyberark cylance elasticsearch envoyproxy f5 fortinet gcp google_workspace googlecloud gsuite haproxy ibmmq icinga iis imperva infoblox iptables juniper kafka kibana logstash microsoft misp mongodb mssql mysql nats netscout nginx o365 okta osquery panw postgresql rabbitmq radware redis santa snort snyk sonicwall sophos squid suricata system tomcat traefik zeek zscaler
  do
    echo "Loading $MODULE"
    docker exec -i so-filebeat filebeat setup modules -pipelines -modules $MODULE -c $FB_MODULE_YML
    sleep 2
  done
else
  exit 0
fi
```

Can I move the files/directories in the filebeats conf directory over from my working test node to have this work?

**Reply by Git-Me-Some-Hub** [reply]
@weslambert - 
I was installing and updating a new Test so-cluster, so I can verify what is supposed to working and what is not on my current main so-cluster. During `soup `I saw it roll through a script for the module setup... Not really sure why it did not work right when I did my upgrades earlier, but my guess it was the issues I had with **so-dockerregistry**. 

That being said, is there a way to just have it roll through those setup options again? Can I change the /etc/soversion and run soup to do all those things again? Maybe a case-by-case script run? 
![image](https://user-images.githubusercontent.com/17930434/127541241-2e24be41-67e0-408e-b3de-63a61c812955.png)

**Reply by Git-Me-Some-Hub** [reply]
Tried to transfer over some files from a working node to a non-working one and still no luck. Must be a bit over my head, I wish there was a capability to run this module setup script all over again and have it work.... 

If not I guess I will look into the work it will take to just start over and transfer all the data over to a new node.

**Comment by jaredatobe** [comment]
This may or may not be related, but to me it looks like there is a section in the logstash init.sls which is not in the filebeat init.sls, and it should be. I've been also trying to get o365 working, but also trying to get encryption working in filebeat like it is in logstash. It's failing because the certs are not being copied into place, which is what's led me to this thread. Here's what i've found:

I was digging through logstash init.sls:
```
sudo less /opt/so/saltstack/default/salt/logstash/init.sls
```
And I can see this:

<a href="https://user-images.githubusercontent.com/87098751/127579836-56812648-8a07-41f1-beb0-cbf8c907359d.png"><img src="https://user-images.githubusercontent.com/87098751/127579836-56812648-8a07-41f1-beb0-cbf8c907359d.png" width="50%"/></a>

This small section copies over the certs I have placed into this ```/etc``` config folder:
```
/opt/so/saltstack/local/salt/logstash/etc/certs
```
...which then become available inside the so-logstash docker at this location:
```
/usr/share/logstash/certs/
```
When this section of the sls runs, it leaves a log trail like this:
```
----------
          ID: lsetcsync
    Function: file.recurse
        Name: /opt/so/conf/logstash/etc
      Result: True
     Comment: Recursively updated /opt/so/conf/logstash/etc
     Started: 14:54:58.025105
    Duration: 609.672 ms
     Changes:   
              ----------
              /opt/so/conf/logstash/etc/certs/Client.crt:
                  ----------
                  diff:
                      New file
                  group:
                      socore
                  mode:
                      0644
                  user:
                      logstash
              /opt/so/conf/logstash/etc/certs/Client.key:
                  ----------
                  diff:
                      New file
                  group:
                      socore
                  mode:
                      0644
                  user:
                      logstash
              /opt/so/conf/logstash/etc/certs/RootCA.crt:
                  ----------
                  diff:
                      New file
                  group:
                      socore
                  mode:
                      0644
                  user:
                      logstash
----------
```
I currently suspect a similar thing is supposed to be happening inside the filebeat init.sls, because in that init.sls, we can see the empty folders /etc and /pki are being created, but they are not referenced thereafter.

I tried fiddling around with a local filebeat init.sls with this small section added (using file permissions/location for filebeat user) but have not been able to get it to move the files into a predictable location, which leads me to believe there is another part of this configuration sync I don't know about yet.

Hope this is helpful. I haven't been able to find where this stuff is documented, so I'm kind of shooting in the dark here, but it does seem related.

**Reply by Git-Me-Some-Hub** [reply]
Thanks for the reply, I had quickly looked at that filebeat init file yesterday or the day before, but it was a bit out of my league and I did see references in there about the module folders that "should" be in the conf folder. However, I will take a deeper look between the working and non-working server.  Since I have this working fine on a test server, I will probably just copy over those folders and see if that changes anything. 

I was not 100% sure, but what issues were you having on O365? I had that working on my test server, as well as Umbrella. I could try and point you in the right direction if you needed it.... Because yes, most documentation is just not there. Happy to help if I can.
